### PR TITLE
suggest `!` for erroneous identifier `not`

### PR DESCRIPTION
**Pull request by zackmdavis**
not_recovery

This supersedes #48858.

r? @petrochenkov

### CODE REVIEW

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -2870,7 +2870,46 @@ impl<'a> Parser<'a> {
...
+                    let found = match self.token {
+                        // FIXME: map keywords to `TokenType::Keyword`
+                        token::Ident(..) => TokenType::Ident,
+                        _ => TokenType::Token(self.token.clone())
+                    };
+                    // Emit the error ...
+                    let mut err = self.diagnostic()
+                        .struct_span_err(self.span,
+                                         &format!("unexpected {} after identifier",
+                                                  found.to_string()));
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: You can use `format!("unexpected {} after identifier", self.this_token_descr())` here.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -2870,7 +2870,46 @@ impl<'a> Parser<'a> {
                 let (span, e) = self.interpolated_or_expr_span(e)?;
                 (lo.to(span), ExprKind::Box(e))
             }
-            _ => return self.parse_dot_or_call_expr(Some(attrs))
+            token::Ident(..) if self.token.is_ident_named("not") => {
+                // `not` is just an ordinary identifier in Rust-the-language,
+                // but as `rustc`-the-compiler, we can issue clever diagnostics
+                // for confused users who really want to say `!`
+                let token_cannot_continue_expr = |t: &token::Token| match *t {
+                    // These tokens can't continue an expression after an ident
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Comment nit: Tokens that can start an expression after `!`, but can't continue an expression after an ident

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -2870,7 +2870,46 @@ impl<'a> Parser<'a> {
...
+                    token::Ident(..) | token::Literal(..) | token::Pound => true,
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Also, it should be easier now to add the `Nt` variants from https://github.com/rust-lang/rust/pull/48858#issuecomment-373946361

**Comment by petrochenkov** [comment]
Meta: opening a new PR moves it into the end of @bors queue, it's usually better to rework an original PR, then it will land sooner.

File: src/libsyntax/parse/parser.rs (hunk 3 above)
Comment (petrochenkov) [reply src/libsyntax/parse/parser.rs]: Also, for `token::Ident` the correct condition is `t is Token(ident) && ident_can_begin_expr(ident)`

**Comment by shepmaster** [comment]
Ping from triage, @zackmdavis ! It's been a week since we heard from you; will you be able to address the review feedback you've received?

**Comment by zackmdavis** [comment]
@shepmaster yes, but next week; very busy lately

**Comment by zackmdavis** [comment]
updated to address feedback, except that I don't know how to write a test example that exercises the nonterminals (and it's more efficient to push this and _admit_ that I don't know if I can't spare the effort to figure it out myself now)

> Meta: opening a new PR moves it into the end of @bors queue, it's usually better to rework an original PR, then it will land sooner.

I _usually_ add commits or force-push existing PRs, but in this case, I don't think there are any lines of code in common between the original submission and the post-reviewer-feedback version; semantically, it feels like a "different PR addressing the same issue" rather than "a revision of the same PR". For something nonessential like this, I'm not that interested in jostling for position in the buildbot queue: that's a zero-sum game among outstanding PRs!

**Comment by TimNN** [comment]
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/362964003). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Resolving deltas: 100% (612068/612068), completed with 4858 local objects.
---
[00:00:41] configure: rust.quiet-tests     := True
---
[00:38:41] ..........................................................................i.........................
[00:38:47] .................i......................................F...........................................
---
[00:39:20] ...............................................................................................i....
[00:39:27] .....................................................................i..............................
---
[00:39:48] ---- [ui] ui/did_you_mean/issue-46836-identifier-not-instead-of-negation.rs stdout ----
[00:39:48]
[00:39:48] error: /checkout/src/test/ui/did_you_mean/issue-46836-identifier-not-instead-of-negation.rs:13: unexpected error: '13:12: 13:19: unexpected `for_you` after identifier'
[00:39:48]
[00:39:48] error: /checkout/src/test/ui/did_you_mean/issue-46836-identifier-not-instead-of-negation.rs:21: unexpected error: '21:15: 21:24: unexpected `the_worst` after identifier'
[00:39:48]
[00:39:48] error: /checkout/src/test/ui/did_you_mean/issue-46836-identifier-not-instead-of-negation.rs:30: unexpected error: '30:9: 30:16: unexpected `println` after identifier'
[00:39:48]
[00:39:48] error: /checkout/src/test/ui/did_you_mean/issue-46836-identifier-not-instead-of-negation.rs:42: unexpected error: '42:27: 42:50: unexpected `be_smothered_out_before` after identifier'
[00:39:48]
[00:39:48] error: /checkout/src/test/ui/did_you_mean/issue-46836-identifier-not-instead-of-negation.rs:13: expected error not found: unexpected identifier after identifier
[00:39:48]
[00:39:48] error: /checkout/src/test/ui/did_you_mean/issue-46836-identifier-not-instead-of-negation.rs:21: expected error not found: unexpected identifier after identifier
[00:39:48]
[00:39:48] error: /checkout/src/test/ui/did_you_mean/issue-46836-identifier-not-instead-of-negation.rs:30: expected error not found: unexpected identifier after identifier
[]         line_num: 30,
[00:39:48]         kind: Some(
[00:39:48]             Error
[00:39:48]         ),
[00:39:48]         msg: "30:9: 30:16: unexpected `println` after identifier"
---
[00:39:48]         msg: "42:27: 42:50: unexpected `be_smothered_out_before` after identifier"
---
[00:39:48]         msg: "unexpected identifier after identifier"
---
[00:39:48]         msg: "unexpected identifier after identifier"
---
[00:39:48]         msg: "unexpected identifier after identifier"
---
[00:39:48]         msg: "unexpected identifier after identifier"
[00:39:48]     }
[00:39:48] ]
[00:39:48]
[00:39:48] thread '[ui] ui/did_you_mean/issue-46836-identifier-not-instead-of-negation.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:1253:13
[00:39:48] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:39:48]
[00:39:48]
[00:39:48] failures:
[00:39:48]     [ui] ui/did_you_mean/issue-46836-identifier-not-instead-of-negation.rs
---
[00:39:48] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-3.9/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zmiri -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zmiri -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "3.9.1\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "alwa3696680 .
```

</details><p></p>

I'm a bot! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**.

**Comment by petrochenkov** [comment]
r=me after making Travis green (updating UI tests)

**Comment by zackmdavis** [comment]
(I wish the check for `//~`-expected errors ran anyway even if the stderr file didn't match; it's too easy to update the latter and forget about the former.)

**Comment by zackmdavis** [comment]
[Three-hour timeout on 64-bit Windows??](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6964/job/tseo6ax4jnh96uoe)

**Pull request merged**